### PR TITLE
Fix client dfget bug and support digest

### PR DESCRIPTION
**Pull request by garfield009**
## Description
fix client dfget bug and support digest

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [ ] Code compiles correctly.
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: internal/dfheaders/headers.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dfheaders
+
+const (
+	Range = "Range"
```
Comment (gaius-qi) [inline internal/dfheaders/headers.go:20]: Range is http header.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: client/dfget/dfget.go [hunk 2]
```diff
@@ -20,160 +20,219 @@ import (
...
-	// todo using progressbar when showBar is true
-	pb := progressbar.DefaultBytes(-1, "Downloading")
-	for {
-		result, err = down.Recv()
-		if err != nil {
-			if de, ok := err.(*dferrors.DfError); ok {
-				logger.Errorf("dragonfly daemon returns error code %d/%s", de.Code, de.Message)
-			} else {
-				logger.Errorf("dragonfly daemon returns error %s", err)
+	go func() {
```
Comment (jim3ma) [inline client/dfget/dfget.go:62]: This anonymous is too long, please extract to a new function.

File: client/dfget/dfget.go [hunk 3]
```diff
@@ -20,160 +20,219 @@ import (
...
 	)
 
-	response, err = source.Download(context.Background(), cfg.URL, hdr)
-	if err != nil {
-		logger.Errorf("download from source error: %s", err)
+	wLog.Info("try to download from source and ignore rate limit")
+	fmt.Println("try to download from source and ignore rate limit")
+
+	dir := filepath.Dir(cfg.Output)
+	_ = fileutils.MkdirAllWithOwner(dir, basic.UserID, basic.UserGroup)
```
Comment (jim3ma) [inline client/dfget/dfget.go:145]: If the user did not has permission to create directories ?

Comment (garfield009) [reply client/dfget/dfget.go:146]: root setuid has permission to create

File: client/dfget/dfget_test.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"d7y.io/dragonfly/v2/client/config"
+	"d7y.io/dragonfly/v2/internal/constants"
+	"d7y.io/dragonfly/v2/internal/idgen"
+	"d7y.io/dragonfly/v2/pkg/source"
+	sourceMock "d7y.io/dragonfly/v2/pkg/source/mock"
```
Comment (jim3ma) [inline client/dfget/dfget_test.go:31]: lower case

File: client/dfget/dfget_test.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	cfg := &config.DfgetConfig{
+		URL:    "http://a.b.c/xx",
+		Output: output,
+		Digest: strings.Join([]string{constants.Sha256Hash, digestutils.Sha256(content)}, ":"),
+	}
+
+	sourceClient.EXPECT().Download(context.Background(), cfg.URL, nil).Return(ioutil.NopCloser(strings.NewReader(content)), nil)
+
+	err = downloadFromSource(context.Background(), cfg, nil)
+	assert.Nil(t, err)
```
Comment (jim3ma) [inline client/dfget/dfget_test.go:58]: Check output content

Comment (garfield009) [reply client/dfget/dfget_test.go:58]: already check digest

**Review by ansinlee — APPROVED** [review]

**Review by 244372610 — APPROVED** [review]

File: client/dfget/dfget.go [hunk 6]
```diff
@@ -20,160 +20,217 @@ import (
...
+			}
+
+			// success
+			if result.Done {
+				if pb != nil {
+					pb.Describe("Downloaded")
+					_ = pb.Close()
+				}
+
+				wLog.Infof("download from daemon success, length:%dByte cost:%dms", result.CompletedLength, time.Now().Sub(start).Milliseconds())
```
Comment (jim3ma) [inline client/dfget/dfget.go:110]: `length:%dByte cost:%dms` -> `length: %dByte cost: %dms`

**Review by gaius-qi — APPROVED** [review]

File: client/dfget/dfget.go [hunk 7]
```diff
@@ -20,160 +20,217 @@ import (
...
+			}
 		}
 	}
-	if err != nil {
-		logger.Errorf("download by dragonfly error: %s", err)
-		return downloadFromSource(cfg, hdr)
+
+	if downError != nil {
+		wLog.Warnf("daemon downloads file error:%v", downError)
+		fmt.Printf("daemon downloads file error:%v\n", downError)
```
Comment (jim3ma) [inline client/dfget/dfget.go:120]: `error:%v` -> `error: %v`

File: client/dfget/dfget.go [hunk 8]
```diff
@@ -20,160 +20,217 @@ import (
...
+	// change file owner
+	if err = os.Chown(target.Name(), basic.UserID, basic.UserGroup); err != nil {
+		return errors.Wrapf(err, "change file owner to uid[%d] gid[%d]", basic.UserID, basic.UserGroup)
+	}
+
+	if err = os.Rename(target.Name(), cfg.Output); err != nil {
 		return err
 	}
+
+	wLog.Infof("download from source success, length:%dByte cost:%dms", written, time.Now().Sub(start).Milliseconds())
```
Comment (jim3ma) [inline client/dfget/dfget.go:177]: `length:%dByte cost:%dms` -> `length: %dByte cost: %dms`

**Pull request merged**